compiler/rustc_passes/src/dead.rs

```diff
@@ -220,7 +242,33 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
                 intravisit::walk_trait_item(self, trait_item);
             }
             Node::ImplItem(impl_item) => {
-                intravisit::walk_impl_item(self, impl_item);
+                if let hir::ImplItemKind::Fn(_, _) = impl_item.kind {
+                    // Get the impl
+                    let impl_maybe_def_id = self.tcx.impl_of_method(impl_item.def_id.to_def_id());
+                    // Get the trait which the impl implements
+                    let trait_maybe_def_id = impl_maybe_def_id
+                        .and_then(|im| self.tcx.impl_trait_ref(im))
+                        .map(|trait_ref| trait_ref.def_id);
+                    // Get the DefIds of Clone and Debug
+                    let clone_maybe_def_id =
+                        self.tcx.lang_items().require(hir::LangItem::Clone).ok();
+                    let debug_maybe_def_id = self.tcx.get_diagnostic_item(sym::debug_trait);
+
+                    if trait_maybe_def_id.is_some()
+                        && (trait_maybe_def_id == clone_maybe_def_id
+                            || trait_maybe_def_id == debug_maybe_def_id)
+                        && impl_item.span.in_derive_expansion()
```

Comment:
```suggestion
                        && self.tcx.has_attr(item.def_id, sym::automatically_derived)
```
this feels better to me, it probably doesn't matter too much though.

Would prevent custom derives of `Clone` to profit from this change :thinking: but maybe that's actually a good thing?

Reply:
This was a conscious choice: I thought that it might be too confusing to actually see the field being read in your code and still get a warning that the "field is never read". Of course, if you'd prefer, we can ignore all impls of (say) `Clone` and `Debug`, not just automatically derived ones.

Reply:
My edit would not ignore all `Clone` impls but ignore only those with the attribute `#[automatically_derived]` which are added by the compiler for the builtin derives.

This change would cause `Clone` impls emitted by other macros, potentially ones which are better at dealing with implied bounds, to not get ignored by the dead code analysis.

Reply:
Ah, I thought you were referring to manual implementations of `Clone`:
```rust
impl Clone for S {
    fn clone(&self) -> Self {
        ...
    }
}
```
I actually included your suggested edit in my last commit; the "conscious choice" I was talking about above referred to manual implementations of `Clone`. Sorry for the confusion!